Sources/Bookmarks/BookmarkUtils.swift

```diff
@@ -0,0 +1,69 @@
+//
+//
+//
+//  Copyright © 2021 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import CoreData
+
+public struct BookmarkUtils {
+        
+    public static func fetchRootFolder(_ context: NSManagedObjectContext) -> BookmarkEntity? {
+        let request = BookmarkEntity.fetchRequest()
+        request.predicate = NSPredicate(format: "%K == %@", #keyPath(BookmarkEntity.uuid), BookmarkEntity.Constants.rootFolderID)
+        request.returnsObjectsAsFaults = false
+        request.fetchLimit = 1
+        
+        return try? context.fetch(request).first
+    }
+    
+    public static func fetchFavoritesFolder(_ context: NSManagedObjectContext) -> BookmarkEntity? {
+        let request = BookmarkEntity.fetchRequest()
+        request.predicate = NSPredicate(format: "%K == %@", #keyPath(BookmarkEntity.uuid), BookmarkEntity.Constants.favoritesFolderID)
+        request.returnsObjectsAsFaults = false
+        request.fetchLimit = 1
+        
+        return try? context.fetch(request).first
+    }
+    
+    public static func prepareFoldersStructure(in context: NSManagedObjectContext) {
+        
+        func insertRootFolder(uuid: String, into context: NSManagedObjectContext) {
+            let folder = BookmarkEntity(entity: BookmarkEntity.entity(in: context),
+                                        insertInto: context)
+            folder.uuid = uuid
+            folder.title = uuid
+            folder.isFolder = true
+        }
+        
+        if fetchRootFolder(context) == nil {
+            insertRootFolder(uuid: BookmarkEntity.Constants.rootFolderID, into: context)
+        }
+        
+        if fetchFavoritesFolder(context) == nil {
+            insertRootFolder(uuid: BookmarkEntity.Constants.favoritesFolderID, into: context)
+        }
+    }
+    
+    public static func fetchBookmark(for url: URL, context: NSManagedObjectContext) -> BookmarkEntity? {
+        let request = BookmarkEntity.fetchRequest()
+        request.predicate = NSPredicate(format: "%K == %@", #keyPath(BookmarkEntity.url), url.absoluteString)
+        request.returnsObjectsAsFaults = false
+        request.fetchLimit = 1
+        
+        return try? context.fetch(request).first
+    }
```

Comment:
This can be confusing if you have multiple bookmarks with the same URL, which is not possible in regular usage but _is_ possible via data import.

I have a website favourited, and when I visit it the browsing menu tells me to add it as a favourite, as I have another instance of that URL somewhere else in the bookmarks store which isn't a favourite.


https://user-images.githubusercontent.com/183774/204688173-9d74b78f-90ae-4e9c-934e-d601e518968d.mp4

Reply:
Bartek has pushed a fix for this, which now mimics the current production app.  You can replicate there by adding a site as a bookmark and a favorite, then dragging the bookmark into favorites so that you have two bookmarks.  The menu shows "favorite" while one exists.

Reply:
Looks good :+1: